validator/src/storage/unsync_block_manager.rs

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::HashMap;
+
+use coretypes::Block;
+use storage::{BlockManager, BlockManagerError, BlockStore, NULL_BLOCK_IDENTIFIER};
+
+struct Anchor {
+    pub blockstore: String,
+    pub ref_count: i64,
+    pub block_id: String,
+}
+
+pub struct UnsyncBlockManager {
+    block_by_block_id: HashMap<String, (Block, i64)>,
+    blockstore_by_name: HashMap<String, Box<BlockStore>>,
+
+    anchors: HashMap<String, Anchor>,
+}
+
+impl UnsyncBlockManager {
+    pub fn new() -> Self {
+        UnsyncBlockManager {
+            block_by_block_id: HashMap::new(),
+            blockstore_by_name: HashMap::new(),
+
+            anchors: HashMap::new(),
+        }
+    }
+
+    fn garbage_collect(&mut self) {
+        let blocks_to_remove: Vec<String> = self.block_by_block_id
+            .iter()
+            .filter(|&(_, &(_, count))| count <= 0)
+            .map(|(block_id, &(_, _))| block_id.to_owned())
+            .collect();
+        blocks_to_remove.iter().for_each(|block_id| {
+            self.block_by_block_id.remove(block_id);
+        });
```

Comment:
I think you can simplify this by removing the collect() just continuing to use the first iterator

Reply:
The problem with that is that `self.block_by_block_id` is being immutably borrowed in the iterator and mutably borrowed in the for_each.